drivers/flash/spi_flash.c

```diff
@@ -0,0 +1,523 @@
+/*
+ * Copyright (c) 2017 Crypta Labs Ltd
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Generic SPI Flash driver for chips with Discoverable Parameters */
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_SPI_LEVEL
+#include <logging/sys_log.h>
+
+#include <errno.h>
+#include <flash.h>
+#include <spi.h>
+#include <misc/util.h>
+
+struct sfdp_header {
+	u32_t	signature;
+	u8_t	minor_ver;
+	u8_t	major_ver;
+	u8_t	nph;
+	u8_t	unused;
+};
+
+struct sfdp_parameter_header {
+	u8_t	id_lsb, minor_ver, major_ver, length, addr[3], id_msb;
+};
+static inline size_t PHDR_ADDR(struct sfdp_parameter_header *phdr)
+{
+	return phdr->addr[0] + (phdr->addr[1]<<8) + (phdr->addr[2]<<16);
+}
+static inline u16_t PHDR_ID(struct sfdp_parameter_header *phdr)
+{
+	return (phdr->id_msb << 8) + phdr->id_lsb;
+}
+
+struct sfdp_basic_flash_parameters {
+	u32_t block_erase_sizes : 2;
+	u32_t write_granularity : 1;
+	u32_t volatile_status_register : 1;
+	u32_t write_enable_opcode_select : 1;
+	u32_t unused1 : 3;
+	u32_t opcode_erase_4k : 8;
+
+	u32_t support_1_1_2_fast_read : 1;
+	u32_t addr_bytes : 2;
+	u32_t support_dtr : 1;
+	u32_t support_1_2_2_fast_read : 1;
+	u32_t support_1_4_4_fast_read : 1;
+	u32_t support_1_1_4_fast_read : 1;
+	u32_t unused2 : 9;
+
+	u32_t density;
+
+	struct {
+		u8_t dummy_clocks : 4;
+		u8_t mode_clocks : 4;
+		u8_t opcode;
+	} read_opcodes[4]; /* 1-4-4, 1-1-4, 1-1-2, 1-2-2 */
+
+	u32_t support_2_2_2_fast_read : 1;
+	u32_t unused3 : 3;
+	u32_t support_4_4_4_fast_read : 1;
+	u32_t unused4 : 27;
+
+	struct {
+		u8_t dummy_clocks : 4;
+		u8_t mode_clocks : 4;
+		u8_t opcode;
+	} read_opcodes_2[4]; /* reserved, 2-2-2, reserved, 4-4-4 */
+
+	struct {
+		u8_t size;
+		u8_t erase_opcode;
+	} sector_sizes[4];
+
+	struct {
+		u8_t mult : 4, time : 4;
+	} sectore_erase_times[4];
+
+	u8_t page_mult : 4;
+	u8_t page_size_shift : 4;
+	u8_t page_program_time : 4;
+	u8_t first_byte_program_time : 4;
+	u8_t additional_byte_program_time : 4;
+	u8_t chip_erase_time : 7;
+	u8_t unused5 : 1;
+
+	u32_t suspend_info;
+
+	u8_t program_resume_opcode;
+	u8_t program_suspend_opcode;
+	u8_t resume_opcode;
+	u8_t suspend_opcode;
+
+	u32_t power_down_info;
+};
+
+#define SFDP_BFPT_ID		0xff00  /* Basic Flash Parameter Table */
+#define SFDP_SECTOR_MAP_ID	0xff81  /* Sector Map Table */
+
+struct spi_flash_data {
+	struct spi_config config;
+	struct k_sem sem;
+
+	u32_t flash_size, page_size;
+	u8_t addr_len;
+	u8_t read_opcode, read_dummy;
+	u8_t write_opcode, write_dummy;
+	u8_t global_unprotect_opcode;
+
+	union {
+		u32_t bfpt[16];
+		struct sfdp_basic_flash_parameters bfp;
+	};
+	u32_t sector_map[16];
+};
+
+#define CMD_IDENTIFY		0x9F
+#define CMD_DISCOVER_PARAMETERS	0x5A
+#define CMD_READ_STATUS		0x05
+#define CMD_READ_MEMORY		0x03
+#define CMD_READ_CONFIG		0x35
+#define CMD_WRITE_STATUS_CONFIG	0x01
+#define CMD_PROGRAM_PAGE	0x02
+#define CMD_WRITE_ENABLE	0x06
+#define CMD_WRITE_DISABLE	0x04
+
+#define STATUS_BUSY		0x01
+
+#define MAX_SFDP_HEADERS	4
+
```

Comment:
Perhaps it would be wise to name it e.g. `drivers/flash/spi_flash_sfdp.[ch]` since we also have other SPI flash drivers in the tree (currently, we have one other SPI flash driver)?